### PR TITLE
Extend tests for ESDT transfers with execution

### PR DESCRIPTION
**Pull request by camilbancioiu**
This PR adds:
* Updated `forwarder-raw.wasm` binary and appropriate updates in the integration tests
* Extra test calls using the `forwarder-raw` smart contract to test the `transferESDTExecute()` EEI method.
* The `TxDataBuilder`
* Updated commit reference to Arwen's `master` branch.

### CODE REVIEW

File: integrationTests/vm/esdt/esdtProcess_test.go [hunk 1]
```diff
@@ -597,81 +574,109 @@ func TestScCallsScWithEsdtIntraShard(t *testing.T) {
...
+	checkAddressHasESDTTokensInt64(t, forwarder, nodes, tokenIdentifier, valueToSendToSc/2)
+	checkAddressHasESDTTokensInt64(t, vault, nodes, tokenIdentifier, valueToSendToSc/2)
 
-	checkNumCallBacks(t, firstScAddress, nodes, 1)
-	checkSavedCallBackData(t, firstScAddress, nodes, 1, "EGLD", big.NewInt(0), vmcommon.Ok, [][]byte{})
+	checkNumCallBacks(t, forwarder, nodes, 1)
+	checkSavedCallBackData(t, forwarder, nodes, 1, "EGLD", big.NewInt(0), vmcommon.Ok, [][]byte{})
 
-	//// call first sc to ask the second one to send it back some esdt
+	// //// call first sc to ask the second one to send it back some esdt
```
Comment (iulianpascalau) [inline integrationTests/vm/esdt/esdtProcess_test.go:634]: Too many //

File: integrationTests/vm/esdt/esdtProcess_test.go [hunk 2]
```diff
@@ -597,81 +574,109 @@ func TestScCallsScWithEsdtIntraShard(t *testing.T) {
...
+	time.Sleep(5 * time.Second)
+	nonce, round = integrationTests.WaitOperationToBeDone(t, nodes, 4, nonce, round, idxProposers)
+	time.Sleep(5 * time.Second)
+
+	tokenIssuerBalance -= valueToTransferWithExecSc
+	checkAddressHasESDTTokensInt64(t, tokenIssuer.OwnAccount.Address, nodes, tokenIdentifier, tokenIssuerBalance)
+	checkAddressHasESDTTokensInt64(t, forwarder, nodes, tokenIdentifier, valueToSendToSc*3/4)
+	checkAddressHasESDTTokensInt64(t, vault, nodes, tokenIdentifier, valueToSendToSc/2)
+
+	//// call first sc to ask the second one to execute a method that transfers ESDT twice, with execution
```
Comment (iulianpascalau) [inline integrationTests/vm/esdt/esdtProcess_test.go:666]: too many //

File: integrationTests/vm/esdt/esdtProcess_test.go [hunk 3]
```diff
@@ -761,14 +766,13 @@ func TestCallbackPaymentEgld(t *testing.T) {
 
 	nonce, round = integrationTests.WaitOperationToBeDone(t, nodes, 4, nonce, round, idxProposers)
 	_, err = nodes[0].AccntState.GetExistingAccount(firstScAddress)
-	assert.Nil(t, err)
+	require.Nil(t, err)
 
+	txData := testVm.NewTxDataBuilder()
 	//// call first sc with esdt, and first sc automatically calls second sc
```
Comment (iulianpascalau) [inline integrationTests/vm/esdt/esdtProcess_test.go:772]: too many // (valid for the rest of the file)

File: integrationTests/vm/esdt/esdtProcess_test.go [hunk 4]
```diff
@@ -879,14 +879,14 @@ func TestScCallsScWithEsdtCrossShard(t *testing.T) {
 
 	nonce, round = integrationTests.WaitOperationToBeDone(t, nodes, 4, nonce, round, idxProposers)
 	_, err = nodes[2].AccntState.GetExistingAccount(firstScAddress)
-	assert.Nil(t, err)
+	require.Nil(t, err)
 
 	//// call first sc with esdt, and first sc automatically calls second sc
 	valueToSendToSc := int64(1000)
 	txData := core.BuiltInFunctionESDTTransfer + "@" +
```
Comment (iulianpascalau) [inline integrationTests/vm/esdt/esdtProcess_test.go:886]: why not use the new fancy txData.New()  ? :D

File: integrationTests/vm/esdt/esdtProcess_test.go [hunk 5]
```diff
@@ -904,7 +904,7 @@ func TestScCallsScWithEsdtCrossShard(t *testing.T) {
 
 	//// call first sc to ask the second one to send it back some esdt
 	valueToRequest := valueToSendToSc / 4
-	txData = "forward_call@" +
+	txData = "forward_async_call@" +
```
Comment (iulianpascalau) [inline integrationTests/vm/esdt/esdtProcess_test.go:907]: txData.New ?

File: integrationTests/vm/esdt/esdtProcess_test.go [hunk 6]
```diff
@@ -1142,10 +1142,143 @@ func TestScCallsScWithEsdtCrossShard_SecondScRefusesPayment(t *testing.T) {
...
+		integrationTests.CreateAndSendTransaction(node, nodes, big.NewInt(0), scAddress, txData, integrationTests.AdditionalGasLimit)
+	}
+
+	time.Sleep(time.Second)
+	_, _ = integrationTests.WaitOperationToBeDone(t, nodes, 4, nonce, round, idxProposers)
+	time.Sleep(time.Second)
+
+	// transferMultiple
+	for _, node := range nodes {
+		txData := "transferEsdtMultiple" + "@" +
```
Comment (iulianpascalau) [inline integrationTests/vm/esdt/esdtProcess_test.go:1271]: txData.New() ?

File: integrationTests/vm/esdt/esdtProcess_test.go [hunk 7]
```diff
@@ -1142,10 +1142,143 @@ func TestScCallsScWithEsdtCrossShard_SecondScRefusesPayment(t *testing.T) {
...
+		integrationTests.CreateAndSendTransaction(node, nodes, big.NewInt(0), scAddress, txData, integrationTests.AdditionalGasLimit)
+	}
+
+	time.Sleep(time.Second)
+	_, _ = integrationTests.WaitOperationToBeDone(t, nodes, 4, nonce, round, idxProposers)
+	time.Sleep(time.Second)
+
+	// transferOne
+	for _, node := range nodes {
+		txData := "transferEsdtOnce" + "@" +
```
Comment (iulianpascalau) [inline integrationTests/vm/esdt/esdtProcess_test.go:1258]: txData.New ?

File: integrationTests/vm/esdt/esdtProcess_test.go [hunk 8]
```diff
@@ -1142,10 +1142,143 @@ func TestScCallsScWithEsdtCrossShard_SecondScRefusesPayment(t *testing.T) {
...
+	)
+
+	nonce, round = integrationTests.WaitOperationToBeDone(t, nodes, 4, nonce, round, idxProposers)
+	_, err := nodes[0].AccntState.GetExistingAccount(scAddress)
+	require.Nil(t, err)
+
+	// call sc with esdt
+	valueToSendToSc := int64(100)
+	for _, node := range nodes {
+		txData := core.BuiltInFunctionESDTTransfer + "@" +
```
Comment (iulianpascalau) [inline integrationTests/vm/esdt/esdtProcess_test.go:1245]: txData.New ?

File: integrationTests/vm/esdt/esdtProcess_test.go [hunk 9]
```diff
@@ -1208,11 +1341,25 @@ func getESDTTokenData(
 
 	tokenKey := []byte(core.ElrondProtectedKeyPrefix + "esdt" + tokenName)
 	esdtData, err := getESDTDataFromKey(userAcc, tokenKey)
-	assert.Nil(t, err)
+	require.Nil(t, err)
 
 	return esdtData
 }
 
+func checkAddressHasESDTTokensInt64(
```
Comment (iulianpascalau) [inline integrationTests/vm/esdt/esdtProcess_test.go:1349]: maybe add a TODO to remove the old `checkAddressHasESDTTokens` function?

File: integrationTests/vm/txDataBuilder.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
+package vm
```
Comment (iulianpascalau) [inline integrationTests/vm/txDataBuilder.go:1]: This is a very useful tool that can be used in other packages other than integrationTests. 
Suggestion to move in /testscommon/txData and call this struct Builder.

Then we would have txData.NewBuilder().....

File: integrationTests/vm/txDataBuilder.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+import (
+	"encoding/hex"
+	"math/big"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+)
+
+// TxDataBuilder constructs a string to be used for transaction arguments
+type TxDataBuilder struct {
```
Comment (iulianpascalau) [inline integrationTests/vm/txDataBuilder.go:11]: we do not need this to be exported, nor its fields as we have a constructor + accessors.

File: integrationTests/vm/txDataBuilder.go [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+		Function:  "",
+		Elements:  make([]string, 0),
+		Separator: "@",
+	}
+}
+
+func (builder *TxDataBuilder) New() *TxDataBuilder {
+	builder.Function = ""
+	builder.Elements = make([]string, 0)
+	return builder
```
Comment (iulianpascalau) [inline integrationTests/vm/txDataBuilder.go:28]: add empty space

File: integrationTests/vm/txDataBuilder.go [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+)
+
+// TxDataBuilder constructs a string to be used for transaction arguments
+type TxDataBuilder struct {
+	Function  string
+	Elements  []string
+	Separator string
+}
+
+func NewTxDataBuilder() *TxDataBuilder {
```
Comment (iulianpascalau) [inline integrationTests/vm/txDataBuilder.go:17]: Missing comments on all exported items

File: integrationTests/vm/txDataBuilder.go [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+	return &TxDataBuilder{
+		Function:  "",
+		Elements:  make([]string, 0),
+		Separator: "@",
+	}
+}
+
+func (builder *TxDataBuilder) New() *TxDataBuilder {
```
Comment (iulianpascalau) [inline integrationTests/vm/txDataBuilder.go:25]: Reset? as to not confuse it with NewTxDataBuilder

File: integrationTests/vm/txDataBuilder.go [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+	return data
+}
+
+func (builder *TxDataBuilder) ToBytes() []byte {
+	return []byte(builder.ToString())
+}
+
+func (builder *TxDataBuilder) Func(function string) *TxDataBuilder {
+	builder.Function = function
+	return builder
```
Comment (iulianpascalau) [inline integrationTests/vm/txDataBuilder.go:46]: add empty space?

File: integrationTests/vm/txDataBuilder.go [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+func (builder *TxDataBuilder) Byte(value byte) *TxDataBuilder {
+	element := hex.EncodeToString([]byte{value})
+	builder.Elements = append(builder.Elements, element)
+	return builder
```
Comment (iulianpascalau) [inline integrationTests/vm/txDataBuilder.go:52]: add empty space on all occurrences where we have something like:
```GO
instruction

return...
```

**Review by iulianpascalau — COMMENTED** [review]
Good implementation! Mostly styling/formatting issues.

File: integrationTests/vm/txDataBuilder.go [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+func (builder *TxDataBuilder) CanPause(prop bool) *TxDataBuilder {
+	return builder.Str("canPause").Bool(prop)
+}
+
+func (builder *TxDataBuilder) CanMint(prop bool) *TxDataBuilder {
+	return builder.Str("canMint").Bool(prop)
+}
+
+func (builder *TxDataBuilder) CanBurn(prop bool) *TxDataBuilder {
```
Comment (sasurobert) [inline integrationTests/vm/txDataBuilder.go:125]: add an IsInterfaceNil

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: integrationTests/vm/txDataBuilder.go (hunk 10 above)
Comment (camilbancioiu) [reply integrationTests/vm/txDataBuilder.go:1]: Moved to package `txDataBuilder`, not `txData`. This avoids ambiguity when we use `txData` as a variable name, which is preferred.

File: integrationTests/vm/txDataBuilder.go (hunk 11 above)
Comment (camilbancioiu) [reply integrationTests/vm/txDataBuilder.go:11]: Made everything private.

File: integrationTests/vm/txDataBuilder.go (hunk 14 above)
Comment (camilbancioiu) [reply integrationTests/vm/txDataBuilder.go:25]: Indeed. Renamed to `Clear()`, though.

File: integrationTests/vm/esdt/esdtProcess_test.go (hunk 1 above)
Comment (camilbancioiu) [reply integrationTests/vm/esdt/esdtProcess_test.go:634]: Removed.

File: integrationTests/vm/esdt/esdtProcess_test.go (hunk 2 above)
Comment (camilbancioiu) [reply integrationTests/vm/esdt/esdtProcess_test.go:666]: Removed.

File: integrationTests/vm/esdt/esdtProcess_test.go (hunk 3 above)
Comment (camilbancioiu) [reply integrationTests/vm/esdt/esdtProcess_test.go:772]: Fixed all occurrences in the file.

File: integrationTests/vm/esdt/esdtProcess_test.go (hunk 5 above)
Comment (camilbancioiu) [reply integrationTests/vm/esdt/esdtProcess_test.go:907]: Added.

File: integrationTests/vm/esdt/esdtProcess_test.go (hunk 4 above)
Comment (camilbancioiu) [reply integrationTests/vm/esdt/esdtProcess_test.go:886]: Added.

File: integrationTests/vm/esdt/esdtProcess_test.go (hunk 6 above)
Comment (camilbancioiu) [reply integrationTests/vm/esdt/esdtProcess_test.go:1271]: Added.

File: integrationTests/vm/esdt/esdtProcess_test.go (hunk 7 above)
Comment (camilbancioiu) [reply integrationTests/vm/esdt/esdtProcess_test.go:1258]: Added.

File: integrationTests/vm/esdt/esdtProcess_test.go (hunk 8 above)
Comment (camilbancioiu) [reply integrationTests/vm/esdt/esdtProcess_test.go:1245]: Added.

File: integrationTests/vm/esdt/esdtProcess_test.go (hunk 9 above)
Comment (camilbancioiu) [reply integrationTests/vm/esdt/esdtProcess_test.go:1349]: Made some simplifications, now there's only one, using `int64`.

File: integrationTests/vm/txDataBuilder.go (hunk 12 above)
Comment (camilbancioiu) [reply integrationTests/vm/txDataBuilder.go:28]: Added.

File: integrationTests/vm/txDataBuilder.go (hunk 13 above)
Comment (camilbancioiu) [reply integrationTests/vm/txDataBuilder.go:17]: Added.

File: integrationTests/vm/txDataBuilder.go (hunk 15 above)
Comment (camilbancioiu) [reply integrationTests/vm/txDataBuilder.go:46]: Added.

File: integrationTests/vm/txDataBuilder.go (hunk 16 above)
Comment (camilbancioiu) [reply integrationTests/vm/txDataBuilder.go:52]: Added.

File: integrationTests/vm/txDataBuilder.go (hunk 17 above)
Comment (camilbancioiu) [reply integrationTests/vm/txDataBuilder.go:125]: Added.

**Review by sasurobert — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**